cvs-debuild.1
.TH CVS-DEBUILD 1 "Debian Utilities" "DEBIAN" \" -*- nroff -*-
.SH NAME
cvs-debuild \- build a Debian package using cvs-buildpackage and debuild
.SH SYNOPSIS
\fBcvs-debuild\fR [\fIdebuild options\fR] [\fIcvs-buildpackage options\fR]
[\fB\-\-lintian-opts\fR \fIlintian options\fR]
[\fB\-\-linda-opts\fR \fIlinda options\fR]
.SH DESCRIPTION
\fBcvs-debuild\fR is a wrapper around \fBcvs-buildpackage\fR to run it
with \fBdebuild\fR as the package-building program.  (This cannot
simply be accomplished using the \fB\-C\fR option of
\fBcvs-buildpackage\fR, as it does not know how to handle all of the
special \fBdebuild\fR options.)
.PP
The program simply stashes the \fBdebuild\fR, \fBlintian\fR and
\fBlinda\fR options, and passes them to \fBdebuild\fR when it is
called by \fBcvs-buildpackage\fR.  All of the standard \fBdebuild\fR
options may be used (as listed below), but note that the root command
specified by any \fB\-\-rootcmd\fR or \fB\-r\fR command-line option
will be passed as an option to \fBcvs-buildpackage\fR.  The first
non-\fBdebuild\fR option detected will signal the start of the
\fBcvs-buildpackage\fR options.
.PP
The selection of the root command is slightly subtle: if there are any
command-line options, these will be used.  If not, then if
\fBcvs-buildpackage\fR is set up to use a default root command, that
will be used.  Finally, if neither of these are the case, then
\fBdebuild\fR will use its procedures to determine an appropriate
command, as described in its documentation.
.PP
See the manpages for \fBdebuild\fR(1) and \fBcvs-buildpackage\fR for
more information about the behaviour of each.
.SH "OPTIONS"
The following are the \fBdebuild\fR options recognised by
\fBcvs-debuild\fR.  All \fBcvs-buildpackage\fR, \fBlintian\fR and
\fBlinda\fR options are simply passed to the appropriate program.  For
explanations of the meanings of these variables, see
\fBdebuild\fR(1).
.TP
.B \-\-no\-conf, \-\-noconf
.TP
.BI \-\-rootcmd= "gain-root-command, " \-r gain-root-command
.TP
.B \-\-preserve\-env
.TP
.BI \-\-preserve\-envvar= "var, " \-e var
.TP
.BI \-\-set\-envvar= var = "value, " \-e var = value
.TP
.B \-\-lintian, \-\-no\-lintian
.TP
.B \-\-linda, \-\-no\-linda
.TP
\fB\-\-ignore-dirname\fR, \fB\-\-check-dirname\fR
These should not be needed, but it is provided nevertheless.
.SH "SEE ALSO"
.BR cvs-buildpackage (1),
.BR debuild (1),
.BR dpkg-buildpackage (1),
.BR lintian (1)
and
.BR linda (1).
.SH AUTHOR
This program was written by Julian Gilbey <jdg@debian.org>.


debian/NEWS
devscripts (2.10.54) UNRELEASED; urgency=low

    manpage-alert now, as originally intended, prints out the full path to the
    binary which is lacking a manpage instead of simply the binary name.

    Before:
      No manual entry for somebinary

    After:
      No manual entry for /usr/bin/somebinary

 -- James Vega <jamessan@debian.org>  Thu, 06 Aug 2009 18:03:41 -0400

devscripts (2.10.24) unstable; urgency=low

    Starting with devscripts 2.10.21, debchange attempts to
    automatically determine whether an NMU changelog should be
    started, by comparing DEBEMAIL and DEBFULLNAME to the
    package's Maintainer and Uploaders fields.

    This behaviour may now be disabled using the DEBCHANGE_AUTO_NMU
    configuration file variable.

 -- Adam D. Barratt <adam@adam-barratt.org.uk>  Mon, 07 Apr 2008 14:38:08 +0100

devscripts (2.10.20) unstable; urgency=low

    devscripts 2.10.19 modified debcommit's --release option to use the
    uncommitted changes (if any) in the changelog to derive the commit
    message.

    This behaviour is now optional and can be enabled using the -R /
    --release-use-changelog command line option or by setting the
    DEBCOMMIT_RELEASE_USE_CHANGELOG configuration option to "yes".

    The default is to use the previous (i.e. pre-2.10.19) behaviour.

 -- Adam D. Barratt <adam@adam-barratt.org.uk>  Wed, 19 Mar 2008 20:31:20 +0000

devscripts (2.10.19) unstable; urgency=low

    In devscripts 2.10.17, dget was modified to unpack source packages by
    default.  The new -d switch or the DGET_UNPACK configuration file option
    may be used to indicate that the unpacking should not occur.

    As linda has been removed from both unstable and testing, support for
    it has been removed from debuild and cvs-debuild.

 -- Adam D. Barratt <adam@adam-barratt.org.uk>  Thu, 13 Mar 2008 19:33:38 +0000


trunk/cvs-debuild.pl
#! /usr/bin/perl -w

# A wrapper for cvs-buildpackage to use debuild, still giving access
# to all of debuild's functionality.

# Copyright 2003, Julian Gilbey <jdg@debian.org>
# 
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA


# We will do simple option processing.  The calling syntax of this
# program is:
#
# cvs-debuild [<debuild options>] [<cvs-buildpackage options>]
#           [--lintian-opts <lintian options>] [--linda-opts <linda options>]
# 
# cvs-debuild will run cvs-buildpackage, using debuild as the
# package-building program, passing the debuild and lintian options to
# it.  For details of these options, and more information on debuild in
# general, refer to debuild(1).

use 5.006;
use strict;
use FileHandle;
use File::Basename;
use File::Temp qw/ tempfile /;
use Fcntl;

my $progname=basename($0);

# Predeclare functions
sub fatal($);

sub usage
{
    print <<"EOF";
  $progname [<debuild options>] [<cvs-buildpackage options>]
             [--lintian-opts <lintian options>] [--linda-opts <linda options>]
  to run cvs-buildpackage using debuild as the package building program

  Accepted debuild options, see debuild(1) or debuild --help for more info:
    --no-conf, --noconf
    --lintian, --no-lintian
    --linda, --no-linda
    --rootcmd=<gain-root-command>, -r<gain-root-command>
    --preserve-envvar=<envvar>, -e<envvar>
    --set-envvar=<envvar>=<value>, -e<envvar>=<value>
    --preserve-env
    --check-dirname-level=<value>, --check-dirname-regex=<regex>
    -d, -D

    --help            display this message
    --version         show version and copyright information
  All cvs-buildpackage options are accepted, as are all lintian/linda options

  Note that any cvs-buildpackage options (command line or configuration file)
  for setting a root command will override any debuild configuration file
  options for this.

Default settings modified by devscripts configuration files:
  (no configuration files are read by $progname)
For information on default debuild settings modified by the
configuration files, run:  debuild --help
EOF
}

sub version
{
    print <<"EOF";
This is $progname, from the Debian devscripts package, version ###VERSION###
This code is copyright 2003 by Julian Gilbey <jdg\@debian.org>,
all rights reserved.
This program comes with ABSOLUTELY NO WARRANTY.
You are free to redistribute this code under the terms of the
GNU General Public License, version 2 or later.
EOF
}

# First check we can execute cvs-buildpackage
unless (system("command -v cvs-buildpackage >/dev/null 2>&1") == 0) {
    fatal "can't run cvs-buildpackage; have you installed it?";
}

# We start by parsing the command line to collect debuild and
# {lintian,linda} options.  We stash them away in temporary files,
# which we will pass to debuild.

my (@debuild_opts, @cvs_opts, @lin_opts);
{
    no locale;
    # debuild opts first
    while (@ARGV) {
	my $arg=shift;
	$arg eq '--help' and usage(), exit 0;
	$arg eq '--version' and version(), exit 0;

	# rootcmd gets passed on to cvs-buildpackage
	if ($arg eq '-r' or $arg eq '--rootcmd') {
	    push @cvs_opts, '-r' . shift;
	    next;
	}
	if ($arg =~ /^(?:-r|--rootcmd=)(.*)$/) {
	    push @cvs_opts, "-r$1";
	    next;
	}

	# other debuild options are stashed
	if ($arg =~ /^--(no-?conf|(no-?)?(lintian|linda))$/) {
	    push @debuild_opts, $arg;
	    next;
	}
	if ($arg =~ /^--(preserve-env|ignore-dirname)$/) {
	    push @debuild_opts, $arg;
	    next;
	}
	if ($arg =~ /^--check-dirname-(level|regex)$/) {
	    push @debuild_opts, $arg, shift;
	    next;
	}
	if ($arg =~ /^--check-dirname-(level|regex)=/) {
	    push @debuild_opts, $arg;
	    next;
	}
	if ($arg =~ /^--(preserve|set)-envvar$/) {
	    push @debuild_opts, $arg, shift;
	    next;
	}
	if ($arg =~ /^--(preserve|set)-envvar=/) {
	    push @debuild_opts, $arg;
	    next;
	}
	# dpkg-buildpackage now has a -e option, so we have to be
	# careful not to confuse the two; their option will always have
	# the form -e<maintainer email> or similar
	if ($arg eq '-e') {
	    push @debuild_opts, $arg, shift;
	    next;
	}
	if ($arg =~ /^-e(\w+(=.*)?)$/) {
	    push @debuild_opts, $arg;
	    next;
	}
	if ($arg eq '-d' or $arg eq '-D') {
	    push @debuild_opts, $arg;
	    next;
	}
	# Anything else matching /^-e/ is a dpkg-buildpackage option,
	# and we've also now considered all debuild options.
	# So now handle cvs-buildpackage options
	unshift @ARGV, $arg;
	last;
    }

    while (@ARGV) {
	my $arg=shift;
	if ($arg eq '-L' or $arg eq '--lintian') {
	    fatal "$arg argument not recognised; use --lintian-opts instead";
	}
	if ($arg =~ /^--lin(tian|da)-opts$/) {
	    push @lin_opts, $arg;
	    last;
	}
	push @cvs_opts, $arg;
    }

    if (@ARGV) {
	push @lin_opts, @ARGV;
    }
}

# So we've now got three arrays, and we'll have to store the debuild
# options in temporary files
my $debuild_cmd='debuild --cvs-debuild';
my ($fhdeb, $fhlin);
if (@debuild_opts) {
    $fhdeb = tempfile("cvspreXXXXXX", UNLINK => 1)
	or fatal "cannot create temporary file: $!";
    fcntl $fhdeb, Fcntl::F_SETFD(), 0
	or fatal "disabling close-on-exec for temporary file: $!";
    print $fhdeb join("\0", @debuild_opts);
    $debuild_cmd .= ' --cvs-debuild-deb /dev/fd/' . fileno($fhdeb);
}
if (@lin_opts) {
    $fhlin = tempfile("cvspreXXXXXX", UNLINK => 1)
	or fatal "cannot create temporary file: $!";
    fcntl $fhlin, Fcntl::F_SETFD(), 0
	or fatal "disabling close-on-exec for temporary file: $!";
    print $fhlin join("\0", @lin_opts);
    $debuild_cmd .= ' --cvs-debuild-lin /dev/fd/' . fileno($fhlin);
}

# Now we can run cvs-buildpackage
my $status = system('cvs-buildpackage', '-C'.$debuild_cmd, @cvs_opts);

if ($status & 255) {
    die "cvs-debuild: cvs-buildpackage terminated abnormally: " .
	sprintf("%#x",$status) . "\n";
} else {
    exit ($status >> 8);
}


sub fatal($) {
    my ($pack,$file,$line);
    ($pack,$file,$line) = caller();
    (my $msg = "$progname: fatal error at line $line:\n@_\n") =~ tr/\0//d;
    $msg =~ s/\n\n$/\n/;
    die $msg;
}
